finality-aleph/src/data_io.rs

```diff
@@ -388,18 +391,19 @@ pub(crate) async fn refresh_best_chain<B, BE, SC, C>(
                     .await
                     .expect("No best chain");
                 if *new_best_header.number() <= max_block_num {
-                    *proposed_block.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
+                    *proposed_block.lock().await = AlephData::new(new_best_header.hash(), *new_best_header.number());
                 } else {
+                    let mut block = proposed_block.lock().await;
```

Comment:
You are holding the lock over a call to `reduce_header_to_num`, which is an expensive operation -- this must be avoided.

Reply:
There are also locks inside `reduce_header_to_num` (client), so I'm not even sure this cannot lead to deadlocks.

Reply:
Well, the thing is that if statement is dependent on lock, and `reduce_header_to_num` happens between if statement and change in locked value. So Im not sure we can avoid it here. Unless there is no problem if we change this value after if statement state changes.

Reply:
The `if condition` depends on `block.number`, but then the next line doesn't.

Reply:
And the `if condition` is evaluated prior to evaluating the `if body`.

Reply:
Ok, I think I see what your concern is. This is the only thread that writes to this variable, so it is safe.

Reply:
The concern is that `block.number` can change during execution of `if body`. The value of `proposed_block` could change (so that other branch of `if` should be selected), and we would override `proposed_block` with a wrong value. This is a bug from a concurrent programming point of view.

That said you are right, that now `proposed_block` is only changed in this place. If we had condition that it can be changed only here, only by this thread then this part is correct. But it comes from `spawn_authority_subtasks` function, so it could be changed in other places in the future and the person changing something here would have to remember that we have a strange lock and in other places we can do only a read lock.
 
But then I looked at the code and it would be really strange if it was changed somewhere else. What do you think about it @timorl ?

Reply:
Eh, I'm not convinced the whole optimization happening in these `if` is such a good idea anyway, so maybe drop at least the external `if`? Then we wouldn't need the (b)lock for anything else than assigning.

Reply:
But the problem is with internal `if`, not with external. Which `if` do you mean by external?

Reply:
Do you mean this optimization?
```
let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), prev_best_number);
```

Reply:
Yeah, the `prev_best_number` part requires some thought, but should be changeable to something reasonable.

Reply:
Ok, so I guess this can be changed to something like that:
```
if *new_best_header.number() <= max_block_num {
    *proposed_block.lock().await = AlephData::new(new_best_header.hash(), *new_best_header.number());
} else {
    // we check if prev_best_header is an ancestor of new_best_header:
    if prev_best_number < max_block_num {
        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
        *proposed_block.lock().await = AlephData::new(reduced_header.hash(), *reduced_header.number());
    } else {
        let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), prev_best_number);
        if reduced_header.hash() != prev_best_hash {
            // the new best block is not a descendant of previous best
            let reduced_header = reduce_header_to_num(client.clone(), new_best_header.clone(), max_block_num);
            *proposed_block.lock().await = AlephData::new(reduced_header.hash(), *reduced_header.number());
        }
    }
}
```
And now there is no problem with locks

Reply:
Yes, this version looks correct to me.